torch/csrc/jit/passes/quantization.cpp

```diff
@@ -2709,6 +2716,66 @@ void ReplicateDeQuant(std::shared_ptr<Graph>& graph) {
   }
 }
 
+void ReplicateChooseQParamsQuant(std::shared_ptr<Graph>& graph) {
+  std::stack<Block*> blocks_to_visit;
+  std::vector<Node*> quant_nodes_to_rewrite;
+  std::vector<Node*> choose_qparam_nodes_to_rewrite;
+  blocks_to_visit.push(graph->block());
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (Node* n : b->nodes()) {
+      if ((n->kind() == Symbol::aten("quantize_per_tensor") ||
+           n->kind() == Symbol::aten("quantize_per_channel")) &&
+          n->output()->uses().size() > 1) {
+        quant_nodes_to_rewrite.push_back(n);
+      }
+      for (Block* subblock : n->blocks()) {
+        blocks_to_visit.push(subblock);
+      }
+    }
+  }
+  for (Node* n : quant_nodes_to_rewrite) {
+    Value* original_val = n->inputs()[0];
+    Value* quantized_val = n->output();
+    Value* dtype = n->inputs()[3];
+    TORCH_CHECK(original_val->uses().size() == 2);
+    for (const Use& use : original_val->uses()) {
+      if (use.user->kind() == Symbol::aten("_choose_qparams_per_tensor")) {
+        choose_qparam_nodes_to_rewrite.push_back(use.user);
+      }
+    }
+    std::vector<Use> uses = quantized_val->uses();
+    for (const Use& use : uses) {
+      auto* user = use.user;
+      WithInsertPoint ins(user);
+      Node* choose_qparam =
+          insertChooseQParamsCall(graph.get(), original_val, user);
+      std::vector<Value*> quant_inputs = {original_val};
+      for (auto& out : choose_qparam->outputs()) {
+        quant_inputs.push_back(out);
+      }
+      quant_inputs.push_back(dtype);
+      Node* quant = graph->create(n->kind(), quant_inputs);
+      quant->output()->setDebugName(original_val->debugName() + ".quant");
+      graph->insertNode(quant);
+      user->replaceInputWith(quantized_val, quant->output());
```

Comment:
I think there should be similar code when you insert choose_qpram - quant - dequant in the beginning, can you refactor this into a function?

Reply:
probably need some refactor of insertQuantDeQuantCall